docs/inference.mdx

```diff
@@ -0,0 +1,134 @@
+
+For purposes of this document, a "request" is the full prompt or input made to the inference service, and a "response" is the text, image, audio, or video returned to the client. The execution of the underlying model is an "inference".
+
+![Inference energy model](images/inference_model.png)
+
+To compute the environment impact of a request, we:
+1. calculate cluster energy use on a continuous basis
+2. use activity logs to calculate the number of concurrent requests at each time
+3. calculate the energy use per second of request duration
+4. model the expected request duration for a given set of request parameters
+
+## Cluster energy use
+
+The cluster is defined by:
+- [Cluster](/cluster) details
+- Cluster location(s) - if hosted in a cloud, which region(s)
+- Cluster throughput (per server/instance if autoscaling)
+
+On a running basis - ideally measured every few seconds - provide:
+- The size of the cluster in nodes (to account for autoscaling)
+- The CPU utilization of each node
+- The utilization of each GPU
+- The power usage of each GPU
+
+As an example, NVIDIA's Triton inference server provides these data through a [metrics service](https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/user_guide/metrics.html).
+
+## Activity logs
+
+For each request, provide:
+
+- Timestamp when computation began
+- Time taken to execute the computation phase of the request
+- Input and output metrics (see below for examples by use case)
+
+### Image metrics
+- Input and output size
+- Output quality
+- LoRA model(s) used
+
+LoRA model training can be modeled using the same methodology as full models.
+
+### Text generation / chat parameters
+- Input tokens
+- Output tokens
+
+## Energy per request duration and complexity
+
+For each request, predict the duration (`predicted_request_duration`) and the number of inferences (`predicted_inferences`). A simple prediction might be:
+```
+predicted_request_duration = avg(duration)
+predicted_inferences = (average output tokens + 1)
+```
+
+Next, calculate the baseline power use of the cluster and spread it across requests by duration:
+
+```
+baseline_energy_per_request_second = E(0,0) / sum(duration)
+```
+
+Calculate the marginal energy per inference:
+```
+marginal_gpu_per_inference = sum(gpu_pct)/sum(inferences)
+marginal_cpu_per_inference = sum(cpu_pct)/sum(inferences)
+
+marginal_energy_per_gpu_pct = E(1,0) - E(0, 0)
+marginal_energy_per_cpu_pct = E(0,1) - E(0, 0)
+
+marginal_energy_per_inference = marginal_gpu_per_inference x marginal_energy_per_gpu_pct +
+                                marginal_cpu_per_inference x marginal_energy_per_cpu_pct
+```
+
+Calculate the energy use for a request:
+```
+usage_energy_per_request = baseline_energy_per_request_second x predicted_request_duration +
```

Comment:
I like this! Main task would be getting the predicted_inferences right (predicted_request_duration is likely directly correlated)